### PR TITLE
feat(core): Add support for passing temperature as a parameter

### LINKED ISSUE #1216: [TAB-433] Add seed / temperature (i64) parameter in code completion request
So the client side have the chance to control randomness and generating a different completion.

Code location:

Request: [https://github.com/TabbyML/tabby/blob/3fdfe652f98b80b5631f45f34cd20ead69111f09/crates/tabby/src/services/completion.rs#L36](https://github.com/TabbyML/tabby/blob/3fdfe652f98b80b5631f45f34cd20ead69111f09/crates/tabby/src/services/completion.rs#L36)

Sampling:
[https://github.com/TabbyML/tabby/blob/3fdfe652f98b80b5631f45f34cd20ead69111f09/crates/llama-cpp-bindings/src/engine.cc#L217](https://github.com/TabbyML/tabby/blob/3fdfe652f98b80b5631f45f34cd20ead69111f09/crates/llama-cpp-bindings/src/engine.cc#L217)

Instead of selecting token with max probs, we need to apply temperature based sampling (with seed): [https://lukesalamone.github.io/posts/what-is-temperature/](https://lukesalamone.github.io/posts/what-is-temperature/)

Might consider reuse [https://github.com/ggerganov/llama.cpp/blob/c37b3474e61d609d43cccc3bde5d559e80e4f5d1/llama.h#L734](https://github.com/ggerganov/llama.cpp/blob/c37b3474e61d609d43cccc3bde5d559e80e4f5d1/llama.h#L734)

<sub>From [SyncLinear.com](https://synclinear.com) | [TAB-433](https://linear.app/tabbyml/issue/TAB-433/add-seed-temperature-i64-parameter-in-code-completion-request-1211)</sub>

### PR DESCRIPTION
**Pull request by boxbeam**
Closes #1216

### CODE REVIEW

File: crates/llama-cpp-bindings/src/engine.cc [hunk 1]
```diff
@@ -110,13 +148,13 @@ class TextInferenceEngineImpl : public TextInferenceEngine {
     llama_batch_free(batch_);
   }
 
-  virtual void add_request(uint32_t request_id, rust::Str text, size_t max_input_length) override {
+  virtual void add_request(uint32_t request_id, rust::Str text, size_t max_input_length, float_t temperature, uint64_t seed) override {
```
Comment (wsxiaoys) [inline crates/llama-cpp-bindings/src/engine.cc:151]: breaking line to make it easier to read

File: crates/llama-cpp-bindings/src/engine.cc [hunk 2]
```diff
@@ -211,11 +249,12 @@ class TextInferenceEngineImpl : public TextInferenceEngine {
         if ((request.i_batch < i) || (request.i_batch >= (i + n_tokens))) {
           continue;
         }
+        srand(request.seed);
 
         int32_t i_batch = request.i_batch - i;
-        auto logits = llama_get_logits_ith(ctx, i_batch);
-        auto next_token = std::distance(logits, std::max_element(logits, logits + n_vocab));
-
+        float* logits = llama_get_logits_ith(ctx, i_batch);
```
Comment (wsxiaoys) [inline crates/llama-cpp-bindings/src/engine.cc:255]:
```suggestion
        const float* logits = llama_get_logits_ith(ctx, i_batch);
```

File: crates/llama-cpp-bindings/src/engine.cc [hunk 3]
```diff
@@ -77,6 +83,38 @@ std::string string_format(const std::string& format, Args ... args)
 	return std::string(buf.get(), buf.get() + size - 1); // We don't want the '\0' inside
 }
 
+float softmax(float* nums, size_t len, float_t temperature) {
```
Comment (wsxiaoys) [inline crates/llama-cpp-bindings/src/engine.cc:86]:
```suggestion
float softmax(const float* nums, size_t len, float temperature) {
```

File: crates/llama-cpp-bindings/src/engine.cc [hunk 4]
```diff
@@ -77,6 +83,38 @@ std::string string_format(const std::string& format, Args ... args)
...
+    nums[i] = std::exp(nums[i] / temperature);
+    sum += nums[i];
+  }
+  for (size_t i = 0; i < len; i++) {
+      nums[i] /= sum;
+  }
+  return sum;
+}
+
+size_t weighted_random(float* nums, size_t len) {
```
Comment (wsxiaoys) [inline crates/llama-cpp-bindings/src/engine.cc:100]: ditto

File: crates/llama-cpp-bindings/src/engine.cc [hunk 5]
```diff
@@ -211,11 +249,12 @@ class TextInferenceEngineImpl : public TextInferenceEngine {
         if ((request.i_batch < i) || (request.i_batch >= (i + n_tokens))) {
           continue;
         }
+        srand(request.seed);
```
Comment (wsxiaoys) [inline crates/llama-cpp-bindings/src/engine.cc:252]: srand itself is not thread safe, let's use std::mt19937 to create a random generator in stack.

Example: https://github.com/ggerganov/llama.cpp/blob/b46757735d30f5c6ed4f20ebeccc684e02d4f3bf/examples/main/main.cpp#L171

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

**Comment by wsxiaoys** [comment]
Please always document the component your pr changed in semantic PR, in this case, it should be 
```
feat(core): ...
```

File: crates/llama-cpp-bindings/src/engine.cc (hunk 3 above)
Comment (boxbeam) [reply crates/llama-cpp-bindings/src/engine.cc:87]: I can't make these const because my `softmax` function modifies the array in-place rather than allocating a new one. Would you rather I allocate one?

File: crates/llama-cpp-bindings/src/engine.cc [hunk 6]
```diff
@@ -211,11 +252,12 @@ class TextInferenceEngineImpl : public TextInferenceEngine {
...
         }
+        std::mt19937 rng(request.seed);
 
         int32_t i_batch = request.i_batch - i;
-        auto logits = llama_get_logits_ith(ctx, i_batch);
-        auto next_token = std::distance(logits, std::max_element(logits, logits + n_vocab));
-
+        float* logits = llama_get_logits_ith(ctx, i_batch);
+        softmax(logits, n_vocab, request.temperature);
+        auto next_token = weighted_random(logits, n_vocab, rng);
```
Comment (wsxiaoys) [inline crates/llama-cpp-bindings/src/engine.cc:260]: pass request.seed as arguments

File: crates/llama-cpp-bindings/src/engine.cc [hunk 7]
```diff
@@ -77,6 +84,38 @@ std::string string_format(const std::string& format, Args ... args)
 	return std::string(buf.get(), buf.get() + size - 1); // We don't want the '\0' inside
 }
 
+float softmax(float* nums, size_t len, float temperature) {
```
Comment (wsxiaoys) [inline crates/llama-cpp-bindings/src/engine.cc:87]:
```suggestion
float compute_softmax_inplace(float* nums, size_t len, float temperature) {
```

File: crates/tabby/src/services/completion.rs [hunk 8]
```diff
@@ -216,7 +227,16 @@ impl CompletionService {
     ) -> Result<CompletionResponse, CompletionError> {
         let completion_id = format!("cmpl-{}", uuid::Uuid::new_v4());
         let language = request.language_or_unknown();
-        let options = Self::text_generation_options(language.as_str());
+        let options = Self::text_generation_options(
+            language.as_str(),
+            request.temperature.unwrap_or(0.1),
+            request.seed.unwrap_or_else(|| {
```
Comment (wsxiaoys) [inline crates/tabby/src/services/completion.rs:233]: could you also add this seed setting logic to `/chat` api? (probably add temperature to that path as well)

Comment (boxbeam) [reply crates/tabby/src/services/completion.rs:233]: Could that be done in a separate PR?

Comment (wsxiaoys) [reply crates/tabby/src/services/completion.rs:233]: sounds good

Comment (boxbeam) [reply crates/tabby/src/services/completion.rs:233]: #1238

File: crates/llama-cpp-bindings/src/engine.cc [hunk 9]
```diff
@@ -77,6 +84,39 @@ std::string string_format(const std::string& format, Args ... args)
...
+  }
+  for (size_t i = 0; i < len; i++) {
+      nums[i] /= sum;
+  }
+  return sum;
+}
+
+size_t weighted_random(const float* nums, size_t len, uint64_t seed) {
+  std::mt19937 rng(seed);
+  float sum = 0;
```
Comment (wsxiaoys) [inline crates/llama-cpp-bindings/src/engine.cc:103]: doesn't softmax above guranteed the `sum` being computed here ends up = `1.0`?

File: crates/llama-cpp-bindings/src/engine.cc [hunk 10]
```diff
@@ -77,6 +84,39 @@ std::string string_format(const std::string& format, Args ... args)
...
+  for (size_t i = 0; i < len; i++) {
+    sum += nums[i];
+  }
+
+  float random = std::uniform_real_distribution<float>(0, sum)(rng);
+  random *= sum;
```
Comment (wsxiaoys) [inline crates/llama-cpp-bindings/src/engine.cc:109]: why need to `*= sum` here?

Comment (boxbeam) [reply crates/llama-cpp-bindings/src/engine.cc:109]: We don't, good catch.

File: crates/llama-cpp-bindings/src/engine.cc (hunk 9 above)
Comment (wsxiaoys) [reply crates/llama-cpp-bindings/src/engine.cc:103]: make sense, while since we compute the sum in `compute_softmax_in_place`, let's just pass it here to avoid recomputing?

File: crates/llama-cpp-bindings/src/engine.cc [hunk 11]
```diff
@@ -77,6 +84,39 @@ std::string string_format(const std::string& format, Args ... args)
...
+  float random = std::uniform_real_distribution<float>(0, sum)(rng);
+  sum = 0;
+  size_t i;
+  for (i = 0; i < len; i++) {
+    sum += nums[i];
+    if (sum >= random) {
+      return i;
+    }
+  }
+  return i;
```
Comment (wsxiaoys) [inline crates/llama-cpp-bindings/src/engine.cc:117]:
```suggestion
  size_t i;
  for (i = 0; i < len; i++) {
    random -= nums[i];
    if (random < 0) {
      return i;
    }
  }
  return i;
```

**Review by wsxiaoys — CHANGES_REQUESTED** [review]
Please have this revisited in after the golden test change for cpu.

File: crates/tabby/src/services/completion.rs
Comment (wsxiaoys) [inline crates/tabby/src/services/completion.rs:1]: Please also add temperature / seed to `chat_completion`, adopting openai's interface.

Comment (boxbeam) [reply crates/tabby/src/services/completion.rs:1]: We had agreed to do this in a separate PR

Comment (wsxiaoys) [reply crates/tabby/src/services/completion.rs:1]: ah - sorry I forgot that discussion. True, it can be done in a separate PR.

**Comment by wsxiaoys** [comment]
cc @icycodes
